packages/firestore/src/core/firestore_client.ts

```diff
@@ -357,6 +367,57 @@ export class FirestoreClient {
     });
   }
 
+  getDocumentFromLocalCache(
+    doc: DocumentReference,
+    observer: Observer<DocumentSnapshot>
+  ): void {
+    this.asyncQueue.schedule(() => {
+      return this.localStore.readDocument(doc._key).then(maybeDoc => {
+        if (maybeDoc instanceof Document) {
+          observer.next(
+            new DocumentSnapshot(doc.firestore, doc._key, maybeDoc, true)
+          );
+        } else {
+          observer.error(
+            new Error(
+              'Failed to get document from cache. (However, this document may ' +
+                "exist on the server. Run again without setting 'source' in " +
+                'the GetOptions to attempt to retrieve the document from the ' +
+                'server.)'
+            )
+          );
+        }
+      });
+    });
+  }
+
+  getDocumentsFromLocalCache(
+    query: Query,
+    observer: Observer<QuerySnapshot>
+  ): void {
+    this.asyncQueue.schedule(() => {
+      return this.localStore.executeQuery(query._query).then(docs => {
+        const remoteKeys: DocumentKeySet = documentKeySet();
+
+        const view = new View(query._query, remoteKeys);
```

Comment:
I was the one that recommended we use the view to compute the snapshot we're sending because they're hard to get right, considering the changes and metadata bits that need to be just so.

I still think that's a good idea: though your idea of defering to the sync engine (or event manager based on which is actually the right place to do this) to pick up the existing snapshot would be worthwhile if it's not too crazy.

It would also have the effect of keeping the FirestoreClient mostly logic free, which is important because it's fiendishly hard to test (today).